README
README file for HTML::Timeline.

See also: Changes.txt.

Warning: WinZip 8.1 and 9.0 both contain an 'accidental' bug which stops
them recognizing POSIX-style directory structures in valid tar files.
You are better off using a reliable tool such as InfoZip:
ftp://ftp.info-zip.org/pub/infozip/

1 Installing from a Unix-like distro
------------------------------------
shell>gunzip HTML-Timeline-1.00-1.00.tgz
shell>tar mxvf HTML-Timeline-1.00-1.00.tar

On Unix-like systems, assuming you have installed Module::Build V 0.25+:

shell>perl Build.PL
shell>./Build
shell>./Build test
shell>./Build install

On MS Windows-like systems, assuming you have installed Module::Build V 0.25+:

shell>perl Build.PL
shell>perl Build
shell>perl Build test
shell>perl Build install

Alternately, without Module::Build, you do this:

Note: 'make' on MS Windows-like systems may be called 'nmake' or 'dmake'.

shell>perl Makefile.PL
shell>make
shell>make test
shell>su              (for Unix-like systems)
shell>make install
shell>exit            (for Unix-like systems)

On all systems:

Run Config.pm through you favourite pod2html translator.

If you are using my fancy-pom2.pl, with its 'default.css' file installed in
/apache2/htdocs/css/, you'd do:

shell>perl fancy-pom2.pl html -css Config.pm > /apache2/htdocs/Config.html

or perhaps something like:

shell>perl fancy-pom2.pl html -css Config.pm > /perl/html/site/lib/HTML/Timeline.html

2 Installing from an ActiveState distro
---------------------------------------
shell>unzip HTML-Timeline-1.00-1.00.zip
shell>ppm install --location=. HTML-Timeline-1.00
shell>del HTML-Timeline-1.00-1.00.ppd
shell>del PPM-HTML-Timeline-1.00-1.00.tar.gz

CHANGES
Revision history for Perl extension HTML::Timeline.

1.08  Unreleased
	- Replace /usr/bin/perl with /usr/bin/env perl.
	- Replace common::sense with use strict and use warnings, to get uninit var warnings.

1.07  Sun Feb 21 12:55:20 2010
	- Remove text 'All rights reserved' (for Debian licensing).
	- Remove POD heads 'Required Modules' and 'Changes'.

1.06  Wed Feb 10 13:52:47 2010
	- Add META.yml. Update MANIFEST.SKIP. Add MANIFEST and MYMETA.yml.

1.05  Fri Nov 13 13:20:00 2009
	- Run dos2unix. This only affects examples/bach.ged.

1.04  Fri Aug 29 14:44:00 2008
	- Download Phil Durbin's patches from github
	- Implement missing_as_table (see docs for details)

1.03  Fri Aug 22 12:23:00 2008
	- Add support for ~/.timelinerc, using Config::IniFiles
	- Rename Changes.txt to CHANGES because search.cpan.org won't display Changes.txt
	- Ensure spouse of root person is included when processing descendents and spouses
	- Fix date-parsing via Gedcom::Date

1.02  Mon Aug 18 13:50:00 2008
	- Switch from Date::Manip as used by ged2xml.pl to Gedcom::Date.
	  Backwards-incompatability, if any, with Date::Manip, is deemed to be a feature
	- Add the -output_dir option. This requires Path::Class
	- Add the -template_dir option. This requires HTML::Template
	- Add the -template_name option
	- Add the -timeline_height option
	- Add the -url_for_xml option
	- Add the -web_page option
	- Rename timeline.html to timeline.tmpl
	- Add user help to the bottom of the HTML output
	- Report people with missing birth dates below the timeline
	- Change default height of the HTML output to 500px
	- Change the hard-coded date in timeline.tmpl to the earliest birth date

1.01  Wed Aug 13 09:27:00 2008
	- Oops. Left Local::Config in t/test.t

1.00  Mon Aug 11 17:56:00 2008
	- Original version


examples/ged2xml.pl
#!/usr/bin/env perl
#
# Note: this version of the script is based on
# commit 8556bb4019b35a285ef7045a33431759b054ee60
# Date:   Sun Aug 3 15:29:58 2008 -0400
#
=head1 NAME

ged2xml.pl - convert GEDCOM files to MIT Simile Timeline XML

=head1 SYNOPSIS

ged2xml.pl [--force] [--xml] [-d5] gedcom_file

 Options:
   -re, --rel        Relationship type: ancestor by default or descendant
   -r, --root        Root person for the tree
   -o, --output      XML output file
   -f, --force       Don't prompt to confirm overwriting XML output file
   -d1, -d5          Set the debug level
   --man             Print the manual page
   -x, --xml         Print the xml to STDOUT

=head1 DESCRIPTION

MIT Simile Timeline ( http://simile.mit.edu/timeline ) loads an XML file containing titles and dates to create a horizontal AJAX timeline of events.  This script generates a compatible XML file from the birth and death dates in a GEDCOM file.

=head1 OPTIONS

=over 2

=item B<-d1, -d2, -d3, -d4, -d5>

Set the debug level.

=item B<-x, --xml>
Print the xml to STDOUT

=back

=head1 BUGS

 I need to at least pull years from unparseable birth dates.

=cut

use strict;
use warnings;
use Getopt::Long;
use Pod::Usage;
Getopt::Long::Configure qw( auto_abbrev auto_version auto_help bundling );
use Readonly;
use List::MoreUtils qw( any );
use Data::Dumper;
# remove private debug module
#use PD::Debug qw( Debug DebugLevel $DEBUG_LEVEL $DEBUG_1 $DEBUG_2 $DEBUG_3 $DEBUG_4 $DEBUG_5 );
use Carp;
use Gedcom;
use Date::Manip;
use XML::Simple;

# used in PD::Debug
## if -zerocolor is given as an argument, disable ANSI color
#if ( any{ /^[-]+z/xmsi } @ARGV ) { $ENV{ANSI_COLORS_DISABLED} = 'yes' }

########################################################################
#
# OPTIONS

my( $HELP, $USAGE, $MAN, $XML, $FORCE, $ALL, $BLOOD_ONLY );

# some defaults
my $ROOT = 'Johann Sebastian Bach';
my $RELATIONSHIP = 'descendents';
my $GEDCOM = 'bach.ged';
my $XML_OUTPUTFILE = 'timeline.xml';

GetOptions(
    'root|r=s'   => \$ROOT, # the person at the root of the tree
    'rel|re=s'   => \$RELATIONSHIP, # ancestors or descendants of person
    'all|a'      => \$ALL, # show everyone in the tree
    'blood|b'    => \$BLOOD_ONLY, # blood relatives only
    'gedcom|g=s' => \$GEDCOM, # the GEDCOM file to parse
    'output|o=s' => \$XML_OUTPUTFILE,
    'force|f'    => \$FORCE,
#    'debug|d:+'  => \$DEBUG_LEVEL,
    'help|h'     => \$HELP,
    'usage|u'    => \$USAGE,
    'man'        => \$MAN,
    'zerocolor|z',
    'xml|x' =>\$XML
) or pod2usage(2);

if ( $HELP  ) { pod2usage(-verbose => 0); }
if ( $USAGE ) { pod2usage(-verbose => 1); }
if ( $MAN   ) { pod2usage(-verbose => 2); }

#Debug( "Debugging level $DEBUG_LEVEL enabled." );

#
########################################################################

########################################################################
# IMPORTANT VARIABLES
#
# XML data to be printed out
my $xml;

#
########################################################################

#my $gedcom_file = "bach.ged";
#DebugLevel( $DEBUG_1, "\$GEDCOM = $GEDCOM\n" );
my $gedcom_file = $GEDCOM;
#DebugLevel( $DEBUG_1, "\$gedcom_file = $gedcom_file\n" );
my $cb; # FIXME: what is this?  delete it?
Readonly my $GEDCOM_VERSION => '5.5';
my $ged = Gedcom->new(grammar_version => $GEDCOM_VERSION,
                      gedcom_file     => $gedcom_file,
                      read_only       => 1,
                      callback        => $cb);
#return unless $ged->validate;
if ( ! $ged->validate) { return };
#my $root = $ged->get_individual("Richard Francis Durbin");
my $root = $ged->get_individual($ROOT);
my $name = _clean_name($root->name);
#print "$name\n";
my $spouse_included;
if ( $BLOOD_ONLY ) {
    $spouse_included = "No";
}
else {
    $spouse_included = "Yes";
}
#DebugLevel( $DEBUG_1, "asked for $ROOT, got $name\n" );
print "The following options have been selected:\n";
print "  Person to start with:    $name\n";
print "  Relationships to list:   $RELATIONSHIP\n";
if ( $RELATIONSHIP eq 'descendents' ) {
    print "  Spouses included:        $spouse_included\n";
}
print "  GEDCOM source file:      $gedcom_file\n";
print "  XML output file:         $XML_OUTPUTFILE\n";
print "Use --help to see how to change these options\n";

my @people;

if ($ALL) {
    # get everyone in GEDCOM file
    print "printing all\n";
    @people = $ged->individuals;
}
else {
    # get people defined by relationship (ancestors, descendents, etc.)
    @people = $root->$RELATIONSHIP;
    my( @spouses, @siblings );
    if ( $RELATIONSHIP eq "descendents" && ! $BLOOD_ONLY ) {
        @spouses = get_spouses($root, @people);
    }
    # get the siblings of the person whose ancestors we want
    if ( $RELATIONSHIP eq "ancestors") {
        @siblings = $root->siblings;
    }
    @people = ($root, @people, @spouses, @siblings );
}

sub get_spouses {
    my @people = @_;
    my @spouses;
    for my $person ( @people ) {
        if ( $person->spouse ) {
            my $spouse = $person->spouse;
            #print $person->name, "\t ", $spouse->name, "\n";
            push( @spouses, $spouse );
        }
    }
    return @spouses;
}

$xml = "<data>\n";
my %seen;
my %notes;
my @missing;
my $missing_message = "The following people would have been included but we don't have a birth date for them: ";
for my $person (@people) {
    #print $person->name, "\n";
    my $name = $person->get_value('name');
    $seen{$person}++;
    if ( $seen{$person} > 1 ) {
        warn "WARNING: $name has already been seen, skipping";
        next;
    }
    #$name =~ s|/||g;
    $name = _clean_name($name);
    #DebugLevel( $DEBUG_1, "\$name = $name\n" );
    #($name) = $name =~ /(\w+)/;
    my $birth_date = $person->get_value('birth date');
    my $death_date = $person->get_value('death date');
    my $summary    = $person->summary;

    # is the date undef?
    # is the date parseable at all?  can we get the year out?
    my $rez = _parse_date($birth_date);
    my $extracted_date;
    # yes there's a birth date, yes, it's valid
    if ($rez) {
        #DebugLevel( $DEBUG_1, "$name birth date: $birth_date" );
        #DebugLevel( $DEBUG_2, "$rez $name" );
        $notes{$name} = ""; # to avoid uninitialized value error
    }
    #yes there's a birth date, no it isn't valid
    elsif ($birth_date) {
        #DebugLevel( $DEBUG_1, "BAD DATE: $summary\n" );
        # FIXME: do something with these notes
        $notes{$name} = "Fuzzy birthdate: $birth_date";
        ($extracted_date) = $birth_date =~ /(\d\d\d\d)/;
    }
    else {
        # skip this person if there is no valid date
        #warn "WARNING: skipping $name as there is no birth date";
        push(@missing, $name);
        next;
    }
    if ($extracted_date ) { $birth_date = $extracted_date; }
    $extracted_date = undef;
    my $valid_death_date = _parse_date($death_date);
    # begin hack for james riley durbin
    # his death date (FEB 1978) is parseable by ParseDate
    # but not similie timeline.  so we'll force it to be
    # invalid so only the year is extracted.
    if ($name eq "James Riley Durbin") {
        warn "WARNING: James Riley Durbin death date hack";
        $valid_death_date = 0;
    }
    # end hack for james riley durbin
    if ( $death_date && ! $valid_death_date ) {
        ($extracted_date) = $death_date =~ /(\d\d\d\d)/;
    }
    if ($extracted_date ) { $death_date = $extracted_date; }

    my $event;
    if ($birth_date && $death_date) {
        my $start = "$birth_date";
        my $end = "$death_date";
        $event .= "<event title=\"$name\" start=\"$start\" end=\"$end\">$notes{$name}</event>\n";
    $xml .= "  $event";
    }
    elsif ($birth_date) {
        my $start = "$birth_date";
        $event .= "<event title=\"$name\" start=\"$start\">$notes{$name}</event>\n";
    $xml .= "  $event";
    }
    #$event .= "/>";
    #$xml .= "  $event\n";
    #$xml .= "  $event";
}
if (@missing) {
    my $today = UnixDate( ParseDate("today"),"%b %e %Y");
    for my $person ( @missing ) {
        $missing_message .= "$person, ";
    }
    chop $missing_message;
    chop $missing_message;
    $xml .= "  <event title=\"Missing\" start=\"$today\">$missing_message</event>\n";
}
$xml .= "</data>\n";
if ( $XML ) { print $xml; }
if (! $FORCE && -e $XML_OUTPUTFILE ) {
    my $answer = 'maybe';
    while ($answer ne 'y' && $answer ne 'n') {
        print "WARNING: Output file $XML_OUTPUTFILE exists.  Overwrite? (y/n): ";
        chomp( $answer = <STDIN>);
        if ($answer eq 'n') {
            print "Exiting.  No changes were made.\n";
            exit;
        }
    }
}
open my $fh, '>', "$XML_OUTPUTFILE" or croak "Couldn't open $XML_OUTPUTFILE: $!";
print {$fh} $xml;
close $fh or croak "Couldn't close $XML_OUTPUTFILE: $!";
print "MIT Simile Timeline XML data source created: $XML_OUTPUTFILE\n";

sub _clean_name {
    my($name) = @_;
    # find forward slashes (/, escaped) everwhere (/g) and remove them
    $name =~ s/\///xmsg;
    return $name;
}

sub _parse_date {
    my( $date )= @_;
    my $result = ParseDate( $date );
    return $result;
}


lib/HTML/Timeline.pm
package HTML::Timeline;

# Author:
#	Ron Savage <ron@savage.net.au>
#
# Note:
#	\t = 4 spaces || die.

use strict;
use warnings;

require 5.005_62;

require Exporter;

# Warning: This list must include format and gedobj, unlike the list in sub new(),
# since those 2 special cases are attributes which are not available to the caller.

use accessors::classic qw/
ancestors
everyone
format
gedcom_file
gedobj
include_spouses
missing_as_table
output_dir
root_person
template_dir
template_name
timeline_height
url_for_xml
verbose
web_page
xml_file
/;
use Carp;
use Gedcom;
use Gedcom::Date;
use HTML::Template;
use Path::Class;

our @ISA = qw(Exporter);

# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

# This allows declaration	use HTML::Timeline ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.
our %EXPORT_TAGS = ( 'all' => [ qw(

) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(

);

our $VERSION = '1.07';

# -----------------------------------------------

sub clean_persons_name
{
	my($self, $name) = @_;

	# Find /s everwhere (/g) and remove them.

	$name =~ s|/||g;

	return $name;

} # End of clean_persons_name.

# -----------------------------------------------

sub generate_xml_file
{
	my($self, $people)   = @_;
	my($missing_message) = 'People excluded because of missing birth dates: ';
	my($todays_date)     = 1900 + (localtime() )[5];

	# Process each person.

	my($birth_date);
	my($death_date);
	my($earliest_date, $extracted_date);
	my(@missing);
	my($name, %notes);
	my($person);
	my($result);
	my(%seen);
	my(@xml);

	push @xml, '<data>';

	for $person (@$people)
	{
		$name = $person -> get_value('name');

		if ($seen{$name})
		{
			$self -> log(sprintf($self -> format(), 'Note', "$name appears twice in the input file") );

			next;
		}

		$seen{$name} = 1;
		$name        = $self -> clean_persons_name($name);
		$birth_date  = $person -> get_value('birth date');
		$death_date  = $person -> get_value('death date');

		# Process birth dates.

		if (Gedcom::Date -> parse($birth_date) )
		{
			$notes{$name} = '';

			if ($earliest_date && ($birth_date < $earliest_date) )
			{
				$earliest_date = $birth_date;
			}
			elsif (! $earliest_date)
			{
				$earliest_date = $birth_date;
			}
		}
		elsif ($birth_date)
		{
			$notes{$name}    = "Fuzzy birthdate: $birth_date";
			($extracted_date = $birth_date) =~ /(\d{4})/;

			if ($extracted_date)
			{
				$birth_date = $extracted_date;
			}

			if ($earliest_date && ($birth_date < $earliest_date) )
			{
				$earliest_date = $birth_date;
			}
			elsif (! $earliest_date)
			{
				$earliest_date = $birth_date;
			}
		}
		else
		{
			push @missing,
			{
				death_date => $death_date,
				name       => $name,
			};

			next;
		}

		# Process death dates.

		if (Gedcom::Date::parse($death_date) )
		{
			# James Riley Durbin's death date (FEB 1978) is parseable by ParseDate
			# but not Similie Timeline, so we only extract the year.

			if ($name eq 'James Riley Durbin')
			{
				($extracted_date = $death_date) =~ /(\d{4})/;

				if ($extracted_date)
				{
					$death_date = $extracted_date;
				}
			}
		}
		elsif ($death_date)
		{
			($extracted_date = $death_date) =~ /(\d{4})/;

			if ($extracted_date)
			{
				$death_date = $extracted_date;
			}
		}

		if ($birth_date && $death_date)
		{
			push @xml, qq|  <event title="$name" start="$birth_date" end="$death_date">$notes{$name}</event>|;
		}
		elsif ($birth_date)
		{
			push @xml, qq|  <event title="$name" start="$birth_date">$notes{$name}</event>|;
		}
	}

	if ( ($self -> missing_as_table() == 0) && ($#missing >= 0) )
	{
		my($missing) = join(', ', map{$$_{'name'} } @missing);

		push @xml, qq|  <event title="Missing" start="$todays_date" end="$todays_date">$missing</event>|;
	}

	push @xml, '</data>';

	# Write timeline.xml.

	my($output_dir)       = $self -> output_dir();
	my($output_file_name) = $self -> xml_file();

	if ($output_dir)
	{
		$output_file_name = file($output_dir, $output_file_name);
	}

	open(OUT, "> $output_file_name") || Carp::croak "Can't open(> $output_file_name): $!";
	print OUT join("\n", @xml), "\n";
	close OUT;

	$self -> log(sprintf($self -> format(), 'Created', $output_file_name) );

	# Write timeline.html.

	my($template)     = HTML::Template -> new(filename => $self -> template_name(), path => $self -> template_dir() );
	my($url_for_xml)  = $self -> url_for_xml();
	$output_file_name = $self -> xml_file();

	if ($url_for_xml)
	{
		$output_file_name = "$url_for_xml/$output_file_name"; # No Path::Class here.
	}

	$template -> param(earliest_date    => $earliest_date);
	$template -> param(missing_as_table => $self -> missing_as_table() );
	$template -> param(timeline_height  => $self -> timeline_height() );
	$template -> param(xml_file_name    => $output_file_name);

	if ($#missing >= 0)
	{
		if ($self -> missing_as_table() == 1)
		{
			$template -> param(missing      => $missing_message);
			$template -> param(missing_loop => [map{ { death_date => $$_{'death_date'}, name => $$_{'name'} } } @missing]);
		}
		else
		{
			$template -> param(todays_date => $todays_date);
		}
	}

	$output_file_name = $self -> web_page();

	if ($output_dir)
	{
		$output_file_name = file($output_dir, $output_file_name);
	}

	open(OUT, "> $output_file_name") || Carp::croak "Can't open(> $output_file_name): $!";
	print OUT $template -> output();
	close OUT;

	$self -> log(sprintf($self -> format(), 'Created', $output_file_name) );

} # End of generate_xml_file.

# -----------------------------------------------

sub get_spouses
{
	my($self, $people) = @_;
	my($spouses)       = [];

	my($person);
	my($spouse);

	for my $person (@$people)
	{
		$spouse = $person -> spouse();

		if ($spouse)
		{
			push @$spouses, $spouse;
		}
	}

	return $spouses;

} # End of get_spouses.

# -----------------------------------------------

sub log
{
	my($self, $message) = @_;

	if ($self -> verbose() )
	{
		print STDERR "$message\n";
	}

} # End of log.

# -----------------------------------------------

sub new
{
	my($class, %arg)    = @_;
	my($self)           = bless({}, $class);
	# Warning: This list must not contain: format or gedobj,
	# since these are attributes not available to the caller.
	my(@options)        = (qw/
ancestors
everyone
gedcom_file
include_spouses
missing_as_table
output_dir
root_person
template_dir
template_name
timeline_height
url_for_xml
verbose
web_page
xml_file
/);

	# Set defaults.

	$self -> ancestors(0);
	$self -> everyone(0);
	$self -> format('%-16s: %s'); # Not in the @options array!
	$self -> gedcom_file('bach.ged');
	$self -> gedobj(''); # Not in the @options array!
	$self -> include_spouses(0);
	$self -> missing_as_table(0);
	$self -> output_dir('');
	$self -> root_person('Johann Sebastian Bach');
	$self -> template_dir('.');
	$self -> template_name('timeline.tmpl');
	$self -> timeline_height(500);
	$self -> url_for_xml('');
	$self -> verbose(0);
	$self -> web_page('timeline.html');
	$self -> xml_file('timeline.xml');

	# Check ~/.timelinerc for more defaults.

	my($resource_file_name) = "$ENV{'HOME'}/.timelinerc";

	if (-e $resource_file_name)
	{
		require "Config/IniFiles.pm";

		my($config)       = Config::IniFiles -> new(-file => $resource_file_name);
		my($section_name) = 'HTML::Timeline';

		if (! $config -> SectionExists($section_name) )
		{
			Carp::croak "Error: Section '$section_name' is missing from $resource_file_name";
		}

		my($option);
		my($value);

		for $option (@options)
		{
			$value = $config -> val($section_name, $option);

			if (defined $value)
			{
				$self -> $option($value);
			}
		}
	}

	# Process user options.

	my($attr_name);

	for $attr_name (@options)
	{
		if (exists($arg{'options'}{$attr_name}) )
		{
			$self -> $attr_name($arg{'options'}{$attr_name});
		}
	}

	if (! -f $self -> gedcom_file() )
	{
		Carp::croak 'Cannot find file: ' . $self -> gedcom_file();
	}

	$self -> gedobj
	(
	 Gedcom -> new
	 (
	  callback        => undef,
	  gedcom_file     => $self -> gedcom_file(),
	  grammar_version => '5.5',
	  read_only       => 1,
	 )
	);

	if (! $self -> gedobj() -> validate() )
	{
		Carp::croak 'Cannot validate file: ' . $self -> gedcom_file();
	}

	$self -> log('Parameters:');

	for $attr_name (@options)
	{
		$self -> log(sprintf($self -> format(), $attr_name, $self -> $attr_name() ) );
	}

	$self -> log('-' x 50);

	return $self;

}	# End of new.

# -----------------------------------------------

sub run
{
	my($self) = @_;

	$self -> log('Processing:');

	my($root_person) = $self -> gedobj() -> get_individual($self -> root_person() );
	my($name)        = $self -> clean_persons_name($root_person -> name() );

	my(@people);

	if ($self -> everyone() == 1)
	{
		@people = $self -> gedobj() -> individuals();
	}
	else
	{
		my($method) = $self -> ancestors() == 1 ? 'ancestors' : 'descendents';
		@people     = $root_person -> $method();

		$self -> log(sprintf($self -> format(), 'Relationship', $method) );

		if ($self -> ancestors() == 0)
		{
			# If descendents are wanted, check for spouses.

			if ($self -> include_spouses() == 1)
			{
				push @people, @{$self -> get_spouses([$root_person, @people])};
			}
		}
		else
		{
			# If ancestors are wanted, check for siblings.

			push @people, $root_person -> siblings();
		}

		unshift @people, $root_person;
	}

	$self -> generate_xml_file(\@people);
	$self -> log('Success');

	return 0;

} # End of run.

# -----------------------------------------------

1;

=head1 NAME

HTML::Timeline - Convert a Gedcom file into a Timeline file

=head1 Synopsis

	shell> bin/timeline.pl -h

=head1 Description

C<HTML::Timeline> is a pure Perl module.

=head1 Distributions

This module is available as a Unix-style distro (*.tgz).

See http://savage.net.au/Perl-modules.html for details.

See http://savage.net.au/Perl-modules/html/installing-a-module.html for
help on unpacking and installing.

=head1 Constructor and initialization

new(...) returns an object of type C<HTML::Timeline>.

This is the class's contructor.

Usage: C<< HTML::Timeline -> new() >>.

This method takes a hashref of options.

Call C<new()> as C<< new({option_1 => value_1, option_2 => value_2, ...}) >>.

See the next section for a discussion of the resource file $HOME/.timelinerc,
which can be used to override the default values for options.

Available options:

=over 4

=item ancestors

If this option is 1, the ancestors of the root_person (see below) are processed.

If this option is 0, their descendents are processed.

The default is 0.

=item everyone

If this option is 1, everyone is processed, and the root_person (see below) is ignored.

If this option is 0, the root_person is processed.

The default is 0.

=item gedcom_file

This takes the name of your input Gedcom file.

The default is bach.ged.

=item include_spouses

If this option is 1, and descendents are processed, spouses are included.

If this option is 0, spouses are ignored.

The default is 0.

=item missing_as_table

If this option is 1, people with missing birthdates are listed under the timeline, in a table.

If this option is 0, such people appear on the timeline, with today's date as their birthdate.

=item output_dir a_dir_name

If this option is used, the output HTML and XML files will be created in this directory.

=item root_person

The name of the person on which to base the timeline.

The default is 'Johann Sebastian Bach'.

=item template_dir a_dir_name

If this option is used, HTML::Template will look in this directory for 'timeline.tmpl'.

If this option is not used, the current directory will be used.

=item template_name a_file_name

If this option is used, HTML::Template will look for a file of this name.

If this option is not used, 'timeline.tmpl' will be used.

=item url_for_xml a_url

If this option is used, it becomes the prefix of the name of the output XML file written into timeline.html.

If this option is not used, no prefix is used.

=item verbose

This takes either a 0 or a 1.

Write more or less progress messages to STDERR.

The default value is 0.

=item web_page a_file_name

If this option is used, it specfies the name of the HTML file to write.

If this option is not used, 'timeline.html' is written.

See the output_dir option for where the file is written.

=item xml_file

The name of your XML output file.

The default value is 'timeline.xml'.

Note: The name of the XML file is embedded in timeline.html, at line 28.
You will need to edit the latter file if you use a different name for your XML output file.

=back

=head1 The resource file $HOME/.timelinerc

The program looks for a file called $HOME/.timelinerc during execution of the constructor.

If this file is present, the module Config::IniFiles is loaded to process it.

If the file is absent, Config::IniFiles does not have to be installed.

This file must contain the section [HTML::Timeline], after which can follow any number
of options, as listed above.

The option names in the file do I<not> start with hyphens.

If the same option appears two or more times, the I<last> appearence is used to set the value
of that option.

The values override the defaults listed above.

These values are, in turn, overridden by the values passed in to the constructor.

This means that command line options passed in to timeline.pl will override the values
found in $HOME/.timelinerc.

Sample file:

	[HTML::Timeline]
	output_dir=/var/www/html

=head1 Method: log($message)

If C<new()> was called as C<< new({verbose => 1}) >>, write the message to STDERR.

If C<new()> was called as C<< new({verbose => 0}) >> (the default), do nothing.

=head1 Method: run()

Do everything.

See C<examples/timeline.pl> for an example of how to call C<run()>.

=head1 See also

The C<Gedcom> module.

=head1 Support

Support is via the Gedcom mailing list.

Subscribe via perl-gedcom-subscribe@perl.org.

=head1 Credits

The MIT Simile Timeline project, and others, are at http://code.google.com/p/simile-widgets/.

Its original home is at http://simile.mit.edu/timeline.

Philip Durbin write the program examples/ged2xml.pl, which Ron Savage converted into a module.

Philip also supplied the files examples/bach.* and examples/timeline.html.

Ron Savage wrote bin/timeline.pl.

examples/timeline.xml is the output of this program, using the default options.

=head1 Author

C<HTML::Timeline> was written by Ron Savage I<E<lt>ron@savage.net.auE<gt>> in 2008.

Home page: http://savage.net.au/index.html

=head1 Copyright

Australian copyright (c) 2008, Ron Savage.

	All Programs of mine are 'OSI Certified Open Source Software';
	you can redistribute them and/or modify them under the terms of
	The Artistic License, a copy of which is available at:
	http://www.opensource.org/licenses/index.html

=cut
